man/ibstat.8
.TH IBSTAT 8 "July 25, 2006" "OpenIB" "OpenIB Diagnostics"

.SH NAME
ibstat \- query basic status of InfiniBand device(s)

.SH SYNOPSIS
.B ibstat
[\-d(ebug)] [\-l(ist_of_cas)] [\-s(hort)] [\-p(ort_list)]
[\-V(ersion)] [\-h] <ca_name> [portnum]

.SH DESCRIPTION
.PP
ibstat is a binary which displays basic information obtained from the local
IB driver. Output includes LID, SMLID, port state, link width active, and port
physical state.

It is similar to the ibstatus utility but implemented as a binary rather
than a script. It has options to list CAs and/or ports and displays more
information than ibstatus.

.SH OPTIONS

.PP
.TP
\fB\-l\fR, \fB\-\-list_of_cas\fR
list all IB devices
.TP
\fB\-s\fR, \fB\-\-short\fR
short output
.TP
\fB\-p\fR, \fB\-\-port_list\fR
show port list
.TP
ca_name
InfiniBand device name
.TP
portnum
port number of InfiniBand device

.SH COMMON OPTIONS

Most OpenIB diagnostics take the following common flags. The exact list of
supported flags per utility can be found in the usage message and can be shown
using the util_name -h syntax.

# Debugging flags
.PP
\-d      raise the IB debugging level.
        May be used several times (-ddd or -d -d -d).
.PP
\-e      show send and receive errors (timeouts and others)
.PP
\-h      show the usage message
.PP
\-v      increase the application verbosity level.
        May be used several times (-vv or -v -v -v)
.PP
\-V      show the version info.

# Addressing flags
.PP
\-D      use directed path address arguments. The path
        is a comma separated list of out ports.
        Examples:
        "0"             # self port
        "0,1,2,1,4"     # out via port 1, then 2, ...
.PP
\-G      use GUID address argument. In most cases, it is the Port GUID.
        Example:
        "0x08f1040023"
.PP
\-s <smlid>      use 'smlid' as the target lid for SM/SA queries.

# Other common flags:
.PP
\-C <ca_name>    use the specified ca_name.
.PP
\-P <ca_port>    use the specified ca_port.
.PP
\-t <timeout_ms> override the default timeout for the solicited mads.

Multiple CA/Multiple Port Support

When no IB device or port is specified, the port to use is selected
by the following criteria:
.PP
1. the first port that is ACTIVE.
.PP
2. if not found, the first port that is UP (physical link up).

If a port and/or CA name is specified, the user request is
attempted to be fulfilled, and will fail if it is not possible.

.SH EXAMPLES

.PP
ibstat            # display status of all ports on all IB devices
.PP
ibstat -l         # list all IB devices
.PP
ibstat -p         # show port guids
.PP
ibstat mthca0 2   # show status of port 2 of 'mthca0'

.SH SEE ALSO
.BR ibstatus (8)

.SH AUTHOR
.TP
Hal Rosenstock
.RI < halr@voltaire.com >


ChangeLog
commit 596aa88065f89906884a821a1bb4eb631abd6af0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:57:58 2009 -0500

    Spin release version 1.2.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0ec532c2b6b92be6adf58849c9e3d215a50d306d
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:52:51 2009 -0500

    Support new device ids: 0x35, 0x36.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f685c8fe7e77e64614d825e563dd9f02a0b1ae16
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:46 2008 -0500

    Spin release 1.2.2
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ebd19d077d331270805aefa9ca00d3661561ba87
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:27 2008 -0500

    Relax the required firmware level.
    
    Turns out that libcxgb3 can really run on either 6.0 or 5.0
    firmware.  So I'm relaxing the requirement...
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 6f7485feb244d8571fcab2292ef92c97bea48df0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue May 13 09:09:35 2008 -0500

    Spin release 1.2.1.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 309916b23ba384a0bd297e7e51b6b89bff357a78
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:45:01 2008 -0500

    Use local stag == 1 to silently drop the p2p read response.
    
    The current code fails if there are read completions -and-
    the p2p read response in the cq when its flushed.
    
    This fix uses the fact that the RTR read is to target stag 1,
    which is not a valid stag.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 67d519c4a9e28e54a9e436404156c9ae7eb60314
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:25:41 2008 -0500

    Handle wrapping when flushing the software sq.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a3dcea751a94ad5e0a9cad6c55a839a9848f327c
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Spin 1.2.0 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit cc2ef741842883233a77e18d65a828c711e581a3
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit 92322b93f48fa98b4fadb736ae49dafda237272a
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Apr 24 14:23:50 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit a8fb6221af4d914f39c1e71f9331f6d428d698c2
Author: Steve Wise (Open Grid Computing) <swise@hosting.openfabrics.org>
Date:   Tue Apr 1 15:30:11 2008 -0700

    Spin 1.1.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit db41ebc10f4cc4c2fe7826589636ab7d4b804218
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 26 08:27:12 2008 -0600

    Provide an empty stub for iwch_query_qp.
    
    This prevents crashes when calling ibv_query_qp due to an expected
    library provided function.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ec4f7f1c6b8beceb9001519bc4e2ee89fbb2e6d0
Author: Dotan Barak <dotanb@dev.mellanox.co.il>
Date:   Tue Feb 26 08:23:36 2008 -0600

    Remove duplicated line.
    
    Signed-off-by: Dotan Barak <dotanb@dev.mellanox.co.il>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 8218482d03898f859e1fa272649fafc64f91f05f
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Feb 12 15:16:03 2008 -0600

    Spin release 1.1.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 1bc6930d8884646ef73db71a95cc32c96c20ba32
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 12 14:46:10 2008 -0600

    libcxgb3: zeroing of wc_flags
    
    The wc_flags field in struct ibv_wc is left uninitialized in
    iwch_poll_cq_one.  User space applications may check this field and
    deterministically perform actions based on the garbage in the field.
    Zeroing this out will prevent this unintended behavior.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 946e1572d38d786038d3bca5a2ec184dcf7636c4
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Fri Jan 25 16:47:01 2008 -0600

    Spin release 1.1.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 94fb771a44212a679d7f83ee69df24e841e0ce08
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Fri Jan 25 13:04:31 2008 -0600

    Fix path for obtaining the firmware version.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 4bbb15965b53cad6273094e95b06895010ca025c
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Wed Jan 23 09:40:28 2008 -0600

    Non-supported functions should return NULL when returning pointers.
    
    Some/Most user space programs will not check for a (void *) to -ENOSYS,
    which can look like a real address until referenced.
    
    This patch converts the uses of (void *) -ENOSYS to NULL.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 10893a7a45e6913483d90023bb90bf9cb2420384
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jan 3 17:19:27 2008 -0600

    Spin release 1.1.2.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a35c74c32d038bd7508ad3cb6c5291c1865d3ba4
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jan 3 17:16:27 2008 -0600

    Fix compiler warnings on 32b.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ef3b9d4f99eb6e18c29b70599edf99d0715f03fb

src/ibstat.c
/*
 * Copyright (c) 2004-2008 Voltaire Inc.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#define _GNU_SOURCE

#if HAVE_CONFIG_H
#  include <config.h>
#endif /* HAVE_CONFIG_H */

#include <inttypes.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <netinet/in.h>

#include <infiniband/umad.h>
#include <infiniband/mad.h>

#include <ibdiag_common.h>

static char *node_type_str[] = {
	"???",
	"CA",
	"Switch",
	"Router",
	"iWARP RNIC"
};

static void
ca_dump(umad_ca_t *ca)
{
	if (!ca->node_type)
		return;
	printf("%s '%s'\n", ((unsigned)ca->node_type <= IB_NODE_MAX ? node_type_str[ca->node_type] : "???"), ca->ca_name);
	printf("\t%s type: %s\n", ((unsigned)ca->node_type <= IB_NODE_MAX ? node_type_str[ca->node_type] : "???"),ca->ca_type);
	printf("\tNumber of ports: %d\n", ca->numports);
	printf("\tFirmware version: %s\n", ca->fw_ver);
	printf("\tHardware version: %s\n", ca->hw_ver);
	printf("\tNode GUID: 0x%016llx\n", (long long unsigned)ntohll(ca->node_guid));
	printf("\tSystem image GUID: 0x%016llx\n", (long long unsigned)ntohll(ca->system_guid));
}

static char *port_state_str[] = {
	"???",
	"Down",
	"Initializing",
	"Armed",
	"Active"
};

static char *port_phy_state_str[] = {
	"No state change",
	"Sleep",
	"Polling",
	"Disabled",
	"PortConfigurationTraining",
	"LinkUp",
	"LinkErrorRecovery",
	"PhyTest"
};

static int
port_dump(umad_port_t *port, int alone)
{
	char *pre = "";
	char *hdrpre = "";

	if (!port)
		return -1;

	if (!alone) {
		pre = "		";
		hdrpre = "	";
	}

	printf("%sPort %d:\n", hdrpre, port->portnum);
	printf("%sState: %s\n", pre, (unsigned)port->state <= 4 ? port_state_str[port->state] : "???");
	printf("%sPhysical state: %s\n", pre, (unsigned)port->state <= 7 ? port_phy_state_str[port->phys_state] : "???");
	printf("%sRate: %d\n", pre, port->rate);
	printf("%sBase lid: %d\n", pre, port->base_lid);
	printf("%sLMC: %d\n", pre, port->lmc);
	printf("%sSM lid: %d\n", pre, port->sm_lid);
	printf("%sCapability mask: 0x%08x\n", pre, ntohl(port->capmask));
	printf("%sPort GUID: 0x%016llx\n", pre, (long long unsigned)ntohll(port->port_guid));
	return 0;
}

static int
ca_stat(char *ca_name, int portnum, int no_ports)
{
	umad_ca_t ca;
	int r;

	if ((r = umad_get_ca(ca_name, &ca)) < 0)
		return r;

	if (!ca.node_type)
		return 0;

	if (!no_ports && portnum >= 0) {
		if (portnum > ca.numports || !ca.ports[portnum]) {
			IBWARN("%s: '%s' has no port number %d - max (%d)",
				((unsigned)ca.node_type <= IB_NODE_MAX ? node_type_str[ca.node_type] : "???"),
				ca_name, portnum, ca.numports);
			return -1;
		}
		printf("%s: '%s'\n", ((unsigned)ca.node_type <= IB_NODE_MAX ? node_type_str[ca.node_type] : "???"), ca.ca_name);
		port_dump(ca.ports[portnum], 1);
		return 0;
	}

	/* print ca header */
	ca_dump(&ca);

	if (no_ports)
		return 0;

	for (portnum = 0; portnum <= ca.numports; portnum++)
		port_dump(ca.ports[portnum], 0);

	return 0;
}

static int
ports_list(char names[][UMAD_CA_NAME_LEN], int n)
{
	uint64_t guids[64];
	int found, ports, i;

	for (i = 0, found = 0; i < n && found < 64; i++) {
		if ((ports = umad_get_ca_portguids(names[i], guids + found, 64 - found)) < 0)
			return -1;
		found += ports;
	}

	for (i = 0; i < found; i++)
		if (guids[i])
			printf("0x%016llx\n", (long long unsigned)ntohll(guids[i]));
	return found;
}

static int list_only, short_format, list_ports;

static int process_opt(void *context, int ch, char *optarg)
{
	switch (ch) {
	case 'l':
		list_only++;
		break;
	case 's':
		short_format++;
		break;
	case 'p':
		list_ports++;
		break;
	default:
		return -1;
	}
	return 0;
}

int main(int argc, char *argv[])
{
	char names[UMAD_MAX_DEVICES][UMAD_CA_NAME_LEN];
	int dev_port = -1;
	int n, i;

	const struct ibdiag_opt opts[] = {
		{ "list_of_cas", 'l', 0, NULL, "list all IB devices" },
		{ "short", 's', 0, NULL, "short output" },
		{ "port_list", 'p', 0, NULL, "show port list" },
		{ 0 }
	};
	char usage_args[] = "<ca_name> [portnum]";
	const char *usage_examples[] = {
		"-l       # list all IB devices",
		"mthca0 2 # stat port 2 of 'mthca0'",
		NULL
	};

	ibdiag_process_opts(argc, argv, NULL, "sDGLCPte", opts, process_opt,
			    usage_args, usage_examples);

	argc -= optind;
	argv += optind;

	if (argc > 1)
		dev_port = strtol(argv[1], 0, 0);

	if (umad_init() < 0)
		IBPANIC("can't init UMAD library");

	if ((n = umad_get_cas_names(names, UMAD_MAX_DEVICES)) < 0)
		IBPANIC("can't list IB device names");

	if (argc) {
		for (i = 0; i < n; i++)
			if (!strncmp(names[i], argv[0], sizeof names[i]))
				break;
		if (i >= n)
			IBPANIC("'%s' IB device can't be found", argv[0]);

		strncpy(names[i], argv[0], sizeof names[i]);
		n = 1;
	}

	if (list_ports) {
		if (ports_list(names, n) < 0)
			IBPANIC("can't list ports");
		return 0;
	}

	if (!list_only && argc) {
		if (ca_stat(argv[0], dev_port, short_format) < 0)
			IBPANIC("stat of IB device '%s' failed", argv[0]);
		return 0;
	}

	for (i = 0; i < n; i++) {
		if (list_only)
			printf("%s\n", names[i]);
		else
			if (ca_stat(names[i], -1, short_format) < 0)
				IBPANIC("stat of IB device '%s' failed", names[i]);
	}

	return 0;
}
